dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java

```diff
@@ -0,0 +1,177 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.hisp.dhis.analytics.event.data.programIndicator;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Date;
+
+import org.hisp.dhis.analytics.AggregationType;
+import org.hisp.dhis.analytics.event.ProgramIndicatorSubqueryBuilder;
+import org.hisp.dhis.commons.util.TextUtils;
+import org.hisp.dhis.program.AnalyticsType;
+import org.hisp.dhis.program.ProgramIndicator;
+import org.hisp.dhis.program.ProgramIndicatorService;
+import org.hisp.dhis.relationship.RelationshipType;
+import org.springframework.stereotype.Component;
+
+import com.google.common.base.Strings;
+
+@Component
+public class DefaultProgramIndicatorSubqueryBuilder
+    implements
+    ProgramIndicatorSubqueryBuilder
+{
+    private final static String ANALYTICS_TABLE_NAME = "analytics";
+
+    private final static String SUBQUERY_TABLE_ALIAS = "subax";
+
+    private final ProgramIndicatorService programIndicatorService;
+
+    public DefaultProgramIndicatorSubqueryBuilder( ProgramIndicatorService programIndicatorService )
+    {
+        checkNotNull( programIndicatorService );
+
+        this.programIndicatorService = programIndicatorService;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getAggregateClauseForProgramIndicator( ProgramIndicator pi, AnalyticsType outerSqlEntity,
+        Date earliestStartDate, Date latestDate )
+    {
+        return _getAggregateClauseForProgramIndicator( pi, null, outerSqlEntity, earliestStartDate, latestDate );
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getAggregateClauseForProgramIndicator( ProgramIndicator programIndicator,
+        RelationshipType relationshipType, AnalyticsType outerSqlEntity, Date earliestStartDate, Date latestDate )
+    {
+        return _getAggregateClauseForProgramIndicator( programIndicator, relationshipType, outerSqlEntity, earliestStartDate, latestDate );
+    }
+
+    private String _getAggregateClauseForProgramIndicator( ProgramIndicator pi, RelationshipType relationshipType,
+        AnalyticsType outerSqlEntity, Date earliestStartDate, Date latestDate )
+    {
+        // Define aggregation function (avg, sum, ...) //
+        String function = TextUtils.emptyIfEqual( pi.getAggregationTypeFallback().getValue(),
+            AggregationType.CUSTOM.getValue() );
+
+        // Get sql construct from Program indicator expression //
+        String aggregateSql = getPrgIndSql( pi.getExpression(), pi, earliestStartDate, latestDate );
+
+        aggregateSql += ")";
```

Comment:
Where is the corresponding "(" coming from here?

Reply:
Added a comment. This parenthesis closes the initial function parenthesis:

`avg( PI query )` <-- this parenthesis